File: web-basics/WebLab3/src/main/java/com/arslanka/backend/models/views/Graph.java
package com.arslanka.backend.models.views;

import lombok.AllArgsConstructor;

@AllArgsConstructor
public class Graph {
    private final Shape topLeft;
    private final Shape topRight;
    private final Shape botRight;
    private final Shape botLeft;

    public Boolean isHit(Double radius, Double y, Double x) {
        if (topLeft != null && x <= 0 && y >= 0) {
            return topLeft.isHit(radius, x, y);
        } else if (topRight != null && x >= 0 && y >= 0) {
            return topRight.isHit(radius, x, y);
        } else if (botRight != null && x >= 0 && y <= 0) {
            return botRight.isHit(radius, x, y);
        } else if (botLeft != null && x <= 0 && y <= 0) {
            return botLeft.isHit(radius, x, y);
        }
        return false;
    }
}

File: web-basics/WebLab4/src/main/kotlin/com/example/arslanka/weblab4/models/views/Graph.kt
package com.example.arslanka.weblab4.models.views

import org.springframework.stereotype.Component
import java.math.BigDecimal

@Component
class Graph(
    private val topLeft: Shape? = Rectangle(Distance.R, Distance.HALF_R), // TODO relocate to prop or env
    private val topRight: Shape? = Circle(Distance.HALF_R),
    private val botRight: Shape? = null,
    private val botLeft: Shape? =Triangle(Distance.R, Distance.R),
) {

    fun isHit(x: BigDecimal, y: BigDecimal, radius: BigDecimal): Boolean {
        if (topLeft != null && x <= BigDecimal.ZERO && y >= BigDecimal.ZERO)
            return topLeft.isHit(radius, x, y)
        else if (topRight != null && x >= BigDecimal.ZERO && y >= BigDecimal.ZERO)
            return topRight.isHit(radius, x, y)
        else if (botRight != null && x >= BigDecimal.ZERO && y <= BigDecimal.ZERO)
            return botRight.isHit(radius, x, y)
        else if (botLeft != null && x <= BigDecimal.ZERO && y <= BigDecimal.ZERO)
            return botLeft.isHit(radius, x, y)
        return false
    }
}
